Add Trusted Bundle Configmap to all namespaces

VaishnaviHire opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Following are the changes introduced - 

1. Introduce a new feature field in DSCI that manages injection of trusted-ca bundle configmap in namepaces
```
spec:
  trustedCABundle:
    managementState: Managed
    customCABundle: ""
```
2. _managementState_ can be **Managed, Unmanaged, Removed.**
    - **Managed**: odh-trusted-ca-bundle configmap will be added to all existing and new namespaces. Any changes to customCABundle will be reflected in the configmap. Configmap state will be maintained by the dscInitialization controller.
    - **Removed**: odh-trusted-ca-bundle configmap if present will be cleaned up by the controller.
    - **Unmanaged**: Configmap will not be managed by the operator, but will not be cleaned from the namespaces.
3. _customCABundle_ field allows admin to add **custom CA bundle** to the configmap.
4. A user if required can explicitly opt-out of the injection of the configmap for a **single namespace** by using annotation security.opendatahub.io/inject-trusted-ca-bundle=false
5. For trustedBundleCA, managementState will set to be _Removed_ for **upgrade** and _Managed_ for **new installs**
 
https://issues.redhat.com/browse/RHOAIENG-2544


## How Has This Been Tested?
1. Install operator using catalogsource quay.io/vhire/opendatahub-operator-catalog:v2.7.0
2. Deploy default DSCI from the DSCInitailization tab for Installed Operators
3.  Verify `odh-trusted-ca-bundle` configmap is created in all non-openshift namespaces
4. Update the value of `.spec.TrustedCABundle.customCABundle` field in DSCI
5. Verify configmaps are updated
6. Add `security.opendatahub.io/inject-trusted-ca-bundle=false` to any user namespace
7.  Verify configmap is removed only for that namespace
8. Update `.spec.TrustedCABundle.managementState` to `Unmanaged` 
9.  Update the cert configmap in any namespace
10. Verify configmap changes are not reverted
11. Update `.spec.TrustedCABundle.managementState` to `Managed` 
12. Verify all changes to the configmap are reverted
13.  Update `.spec.TrustedCABundle.managementState` to `Removed` 
14. Verify all the cert configmaps are removed


## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

zdtsw commented on controllers/dscinitialization/utils.go:
```diff
@@ -426,6 +428,64 @@ func (r *DSCInitializationReconciler) createOdhCommonConfigMap(ctx context.Conte
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// Set Controller reference
+			err = ctrl.SetControllerReference(dscInit, foundConfigMap, r.Scheme)
```
so we will need to have a DSCI CR in order to keep this configmap , in order make cert work?

VaishnaviHire replied:
Yes, I think for any valid workflow of ODH operator,, we require DSCI instance to be created.

Alternatively, we would need a separate controller then if we want to watch over namespaces irrespective of DSC/DSCI resources. The reason I added this is DSCI controller is, since we would also want to update the configmap on changes in the `TrustedCABundle` field.

Let me know if this approach looks fine to you

zdtsw commented on controllers/dscinitialization/utils.go:
```diff
@@ -34,6 +35,7 @@ var (
 // - Odh specific labels
 // - Pod security labels for baseline permissions
 // - ConfigMap  'odh-common-config'
+// - ConfigMap  'odh-trusted-ca-bundle' -- Certificates for the cluster trusted CA Cert Bundle
```
this should be for function createOdhTrustedCABundleConfigMap?

zdtsw commented on controllers/dscinitialization/utils.go:
```diff
@@ -453,3 +513,26 @@ func (r *DSCInitializationReconciler) createUserGroup(ctx context.Context, dscIn
+		return nil
+	}
+
+	for _, ns := range namespaceList.Items {
```
do we want to rule out "default" namespace too?

VaishnaviHire replied:
Yes I will update the condition

LaVLaS commented:
Looks great so far.  One issues is `6. Verify configmap is deleted from all namespaces.`.  Since the expectation is this ConfigMap could/will be mounted to a running Pod, we can't delete it if the DSCI is removed.  We should reconcile these configMaps if there is a changed to the `DSCI .spec.TrustedCABundle` but we should leave it in place unless we want to check every Pod in each namespace to see if the ConfigMap can be safely deleted.

zdtsw commented:
> Looks great so far. One issues is `6. Verify configmap is deleted from all namespaces.`. Since the expectation is this ConfigMap could/will be mounted to a running Pod, we can't delete it if the DSCI is removed. We should reconcile these configMaps if there is a changed to the `DSCI .spec.TrustedCABundle` but we should leave it in place unless we want to check every Pod in each namespace to see if the ConfigMap can be safely deleted.

This gets me to think:
if the `DSCI .spec.TrustedCABundle` gets updated during pods' runtime, how pods would know what the  new` .data`is?
should component teams consider this case?

bartoszmajsak reviewed (commented):
I went through the code ...I have some questions.... "aaaah! this annoying Bartosz again!" :D

bartoszmajsak commented on controllers/dscinitialization/dscinitialization_controller.go:
```diff
@@ -163,6 +164,13 @@ func (r *DSCInitializationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	}
 
+	// Verify odh-trusted-ca-bundle Configmap is created for all the namespaces
+	err = r.addTrustedCABundleConfigMap(ctx, instance)
```
Maybe path worth taking is something different:

- when we start the controller do the bulk update of existing namespaces (matching some criteria)
- then reconcile only on creation of the new namespace

VaishnaviHire replied:
The reason I added it here is, that we would need to reconcile the configmap if there are any changes to the `spec.trustedCABundle` field in DSCI.

Also note, that this is a short-term solution until we have an identifier for Data Science Project namespaces.

bartoszmajsak replied:
That's why I put _even unrelated_ changes in my original comment :) I know that we need to update the bundle. It's intentional that we do not have access to old and new instances in the main reconcile func, so I am not suggesting hacking our way out of it.

What I was wondering is: 
- we know the previous value if it's stored in the cfgmap in e.g. appNs
- we can check if it's different now and only then trigger reconcile for all ns if needed.

> Also note, that this is a short-term solution until we have an identifier for Data Science Project namespaces.

I am well aware of this limitation :) But in Eastern Europe we have this saying: "Nothing is more permanent than a temporary solution" :) And I learned this the hard way.

So my line of thinking is to make sure it's as good as it can be considering the circumstances.

VaishnaviHire commented:
> Looks great so far. One issues is `6. Verify configmap is deleted from all namespaces.`. Since the expectation is this ConfigMap could/will be mounted to a running Pod, we can't delete it if the DSCI is removed. We should reconcile these configMaps if there is a changed to the `DSCI .spec.TrustedCABundle` but we should leave it in place unless we want to check every Pod in each namespace to see if the ConfigMap can be safely deleted.

Okay, I will remove ownerreference of DSCI from the configmap. Just note there will be no other way to delete the configmap automaically, users will have to delete those if needed. However having cert configmaps is common for OpenShift namespaces.

VaishnaviHire commented on controllers/dscinitialization/utils.go:
```diff
@@ -453,3 +514,25 @@ func (r *DSCInitializationReconciler) createUserGroup(ctx context.Context, dscIn
+	}
+
+	for _, ns := range namespaceList.Items {
+		if !strings.HasPrefix(ns.Name, "openshift-") && !strings.HasPrefix(ns.Name, "kube-") &&
+			ns.Name != "default" && ns.Name != "openshift" {
```
Yep will update it

VaishnaviHire commented on controllers/dscinitialization/utils.go:
```diff
@@ -426,6 +427,66 @@ func (r *DSCInitializationReconciler) createOdhCommonConfigMap(ctx context.Conte
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// Set Controller reference
+			err = ctrl.SetControllerReference(dscInit, desiredConfigMap, r.Scheme)
+			if err != nil {
+				r.Log.Error(err, "Unable to add OwnerReference to the", caConfigMapName, "ConfigMap")
+				return err
+			}
```
Agree, I will remove the ownerref

VaishnaviHire commented on controllers/dscinitialization/utils.go:
```diff
@@ -426,6 +427,66 @@ func (r *DSCInitializationReconciler) createOdhCommonConfigMap(ctx context.Conte
 
+// createOdhTrustedCABundleConfigMap creates a configMap  'odh-trusted-ca-bundle' -- Certificates for the cluster
+// trusted CA Cert Bundle.
+func (r *DSCInitializationReconciler) createOdhTrustedCABundleConfigMap(ctx context.Context, namespace string, dscInit *dsci.DSCInitialization) error {
```
Yes, we can do that

VaishnaviHire commented on controllers/dscinitialization/dscinitialization_controller.go:
```diff
@@ -391,6 +406,14 @@ func (r *DSCInitializationReconciler) watchDSCResource(_ client.Object) []reconc
+	if !strings.HasPrefix(a.GetName(), "openshift-") && !strings.HasPrefix(a.GetName(), "kube-") &&
+		a.GetName() != "default" && a.GetName() != "openshift" {
+		r.Log.Info("Found a user namespace, start reconcile")
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{Name: a.GetName()}}}
```
Yes, probably it would be better to add reference to the DSCI instead of namespace.

So there are two things here that will trigger the configmap update:
1. Creation of new namespace
2. Update in DSCI `.spec.trustedCABundle` field.

I believe having a separate controller will not cover update in DSCI.
Let me know if I am missing something here

bartoszmajsak replied:
I agree it will not, but we have captured it in another thread https://github.com/opendatahub-io/opendatahub-operator/pull/848#discussion_r1481530886

I still believe that would make sense to separate.

VaishnaviHire replied:
Updated to introduce a new controller

VaishnaviHire commented:
> > Looks great so far. One issues is `6. Verify configmap is deleted from all namespaces.`. Since the expectation is this ConfigMap could/will be mounted to a running Pod, we can't delete it if the DSCI is removed. We should reconcile these configMaps if there is a changed to the `DSCI .spec.TrustedCABundle` but we should leave it in place unless we want to check every Pod in each namespace to see if the ConfigMap can be safely deleted.
> 
> This gets me to think: if the `DSCI .spec.TrustedCABundle` gets updated during pods' runtime, how pods would know what the new` .data`is? should component teams consider this case?

Yes correct, component controllers will have to add a `Watch` for this configmap.

bartoszmajsak commented on controllers/dscinitialization/dscinitialization_controller.go:
```diff
@@ -391,6 +406,14 @@ func (r *DSCInitializationReconciler) watchDSCResource(_ client.Object) []reconc
+}
 
+func (r *DSCInitializationReconciler) watchNamespaceResource(a client.Object) []reconcile.Request {
+	if !strings.HasPrefix(a.GetName(), "openshift-") && !strings.HasPrefix(a.GetName(), "kube-") &&
```
Thinking more about it, maybe what we need is (for now) opt-out annotation `security.opendatahub.io/inject-trusted-ca-bundle`, which, if found and set to `false`, will not inject the configmap to a given namespace?

bartoszmajsak replied:
WDYT @VaishnaviHire @LaVLaS?

zdtsw commented on pkg/trustedcabundle/trustedcabundle.go:
```diff
@@ -0,0 +1,157 @@
+// createOdhTrustedCABundleConfigMap creates a configMap  'odh-trusted-ca-bundle' -- Certificates for the cluster
+// trusted CA Cert Bundle.
+func CreateOdhTrustedCABundleConfigMap(ctx context.Context, cli client.Client, namespace string, dscInit *dsci.DSCInitialization) error {
+	caConfigMapName := "odh-trusted-ca-bundle"
+	caDataFieldName := "odh-ca-bundle.crt"
```
should just use CAConfigMapName and CADataFieldName ?

zdtsw commented on pkg/trustedcabundle/trustedcabundle.go:
```diff
@@ -0,0 +1,153 @@
+			Name:      CAConfigMapName,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/part-of": "opendatahub-operator",
```
can it be something e.g "app.kubernetes.io/part-of": "trusted-cabundle"
so it can be used for both upstream and downstream ?

zdtsw replied:
> > can it be something e.g "app.kubernetes.io/part-of": "trusted-cabundle"
> 
> But what would it mean? `trusted-cabundle` is not a higher-level app. At least that's how I would understand recommendations here https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels
> 
> Maybe we need a way to parametrize the operator name used in labels based on the distribution?

i could not think of something better than the original value of "opendatahub-operator"
on the other hand, resource like DSC and DSCI CR are labelled by bundle's alm-example or initialization-resource, If here we hardcode it, unless we have different code in downstream, it will be used this as you said for differnt distributions.

VaishnaviHire replied:
Keeping  this as "opendatahub-operator" for now

VaishnaviHire commented:
Hi @LaVLaS @zdtsw @bartoszmajsak I have addressed all the comments. Following are the changes introduced - 

1. Dedicated controller that -
     i. adds `odh-trusted-ca-bundle` configmap to new namespaces
     ii. Maintains the state of configmap if user explicitly modifies/deletes it
     iii. Add support for `security.opendatahub.io/inject-trusted-ca-bundle=false` annotation if they want to opt-out of certificate injection in their namespace.

2. DSCInitialization controller responsible for watching for changes in the `trustedCABundle` fields, and only then updating all namespace.
3. Configmap will not be removed on DSCI deletion. 
4. Configmap can only be removed if users explicitly set the above annotation.

bartoszmajsak commented on controllers/dscinitialization/dscinitialization_test.go:
```diff
@@ -102,6 +104,16 @@ var _ = Describe("DataScienceCluster initialization", func() {
 			Expect(foundConfigMap.Data).To(Equal(expectedConfigmapData))
 		})
+
+		It("Should create the trusted CA Bundle configmap", func() {
+			// then
+			foundConfigMap := &corev1.ConfigMap{}
+			Eventually(objectExists(caBundleConfigmapName, applicationNamespace, foundConfigMap), timeout, interval).Should(BeTrue())
+			Expect(foundConfigMap.Name).To(Equal(caBundleConfigmapName))
+			Expect(foundConfigMap.Namespace).To(Equal(applicationNamespace))
+			expectedConfigmapData := map[string]string{caBundleDataField: ""}
+			Expect(foundConfigMap.Data).To(Equal(expectedConfigmapData))
+		})
```
small suggestion that improves readability

```suggestion
		It("Should create the trusted CA Bundle configmap", func() {
			// then
			foundConfigMap := &corev1.ConfigMap{}
			Eventually(objectExists(caBundleConfigmapName, applicationNamespace, foundConfigMap)).
				WithTimeout(timeout).
				WithPolling(interval).
				Should(BeTrue())
			
			Expect(foundConfigMap.Name).To(Equal(caBundleConfigmapName))
			Expect(foundConfigMap.Namespace).To(Equal(applicationNamespace))
			Expect(foundConfigMap.Data).To(HaveKeyWithValue(caBundleDataField, ""))
		})
```

bartoszmajsak reviewed (approved):
Thanks for taking my suggestions into consideration! I added a few more, of course! But this time easily applicable from GH if you think they make sense.

bartoszmajsak commented on apis/dscinitialization/v1/dscinitialization_types.go:
```diff
@@ -45,9 +45,15 @@ type DSCInitializationSpec struct {
+	// ConfigMap .data.odh-ca-bundle.crt within the DSCInitializtion.spec.ApplicationsNamespace.
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,order=4
+	// +optional
+	TrustedCABundle string `json:"trustedCABundle,omitempty"`
```
we could make it a pointer, then it will really be optional in YAML

VaishnaviHire commented on controllers/certconfigmapgenerator/certconfigmapgenerator_controller.go:
```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2023.
```
Yeah I think these are auto-generated when we create an api using `operator-sdk`. Yep, we can create the issue. We should be able to go through all other redundant usages.

zdtsw commented on pkg/trustedcabundle/trustedcabundle.go:
```diff
@@ -0,0 +1,153 @@
+
+	for _, ns := range namespaceList.Items {
+		if !strings.HasPrefix(ns.Name, "openshift-") && !strings.HasPrefix(ns.Name, "kube-") &&
+			ns.Name != "default" && ns.Name != "openshift" {
```
one last piece, from https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/concepts/projects/utils.ts 
"system" is opt-out, should we exclude it here too?

VaishnaviHire replied:
Not adding `system` to the excluded namespaces, as it is not included in ocp default namespaces.

zdtsw commented on controllers/certconfigmapgenerator/certconfigmapgenerator_controller.go:
```diff
@@ -0,0 +1,140 @@
+
+	// Verify odh-trusted-ca-bundle Configmap is created for the new namespace
+	if trustedcabundle.ShouldInjectTrustedBundle(userNamespace) {
+		err = trustedcabundle.CreateOdhTrustedCABundleConfigMap(ctx, r.Client, req.Namespace, dsciInstance)
```
can just pass the string,  no need dsci CR, 
most of the new functions passing *dsci.DSCInitialization is only to check/set/use `dscInit.Spec.TrustedCABundle.CustomCABundle ` value 
e.g 
```suggestion
       trustCAData :=  dscInit.Spec.TrustedCABundle.CustomCABundle
		err = trustedcabundle.CreateOdhTrustedCABundleConfigMap(ctx, r.Client, req.Namespace, trustCAData)
```

etirelli commented on controllers/certconfigmapgenerator/certconfigmapgenerator_controller.go:
```diff
@@ -0,0 +1,140 @@
+			r.Log.Error(err, "error deleting existing configmap from namespace", "name", trustedcabundle.CAConfigMapName, "namespace", req.Namespace)
+			return reconcile.Result{}, err
+		}
+	}
```
minor nitpick: in case of success, also return at this point to avoid possible future bugs? Or make it an "else if".

etirelli commented on pkg/trustedcabundle/trustedcabundle.go:
```diff
@@ -0,0 +1,151 @@
+		ns := &namespaceList.Items[i]
+		if ShouldInjectTrustedBundle(ns) {
+			if err := CreateOdhTrustedCABundleConfigMap(ctx, cli, ns.Name, dscInit); err != nil {
+				return err
```
question: should we return if the operation fails for a single namespace? or should we try to add to all namespaces and collect a list of errors returning at the end? I am not sure about it.

etirelli reviewed (approved):
LGTM, I left some minor nitpick comments.

shalberd commented:
@VaishnaviHire is this in addition to the mechanism of central openshift? https://docs.openshift.com/container-platform/4.12/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki

as it's currently created by and used by odh notebook controller?

shalberd commented on pkg/trustedcabundle/trustedcabundle.go:
```diff
@@ -0,0 +1,151 @@
+
+// createOdhTrustedCABundleConfigMap creates a configMap  'odh-trusted-ca-bundle' -- Certificates for the cluster
+// trusted CA Cert Bundle.
+func CreateOdhTrustedCABundleConfigMap(ctx context.Context, cli client.Client, namespace string, customCAData string) error {
```
https://docs.openshift.com/container-platform/4.12/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki

the configmap, will it use the openshift certificate injection mechanism? In our corporate environments, all additonal CAs and self-signed trusted certificates are specified centrally at Openshift level and then used by odh notebook controller and in workbenches in this way:

https://github.com/opendatahub-io/kubeflow/pull/43/files#diff-447c1a1ddad4e46669c4371d0d9714dad4a3368c5ccf2292b356e3b5c7441ce1R198

VaishnaviHire replied:
Yes that is correct. Notebook controller will also be updated to reference this configmap. 

cc: @harshad16

shalberd commented on pkg/trustedcabundle/trustedcabundle.go:
```diff
@@ -0,0 +1,151 @@
+				"app.kubernetes.io/part-of": "opendatahub-operator",
+				// Label required for the Cluster Network Operator(CNO) to inject the cluster trusted CA bundle
+				// into .data["ca-bundle.crt"]
+				"config.openshift.io/inject-trusted-cabundle": "true",
```
neat! yeah, so cool

shalberd commented:
how pods would know what the new .data is? should component teams consider this case?

running pods keep the previous value of .data for as long as they run, until restart. As Vaishnavi said, there are mechanisms to add the new content to the running pod without restart, but that is not openshift or k8s standard.

As to whether configmaps can be deleted when they are referenced in running pod volume mounts: I don't think they can, but in any case, i'd be careful about deleting the configmap. Updating its content is fine.

I think the content does not auto-update, though, configmap would need to be re-created / re-applied. Case in point: our cluster admins add a new trusted CA to the central proxy / ca-bundle config ... I don't think that would just auto-update itself in an existing configmap with https://docs.openshift.com/container-platform/4.12/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki. I have never tried that out ...

zdtsw commented:
some tests:
- by adding "security.opendatahub.io/inject-trusted-ca-bundle:false" removed CM odh-trusted-ca-bundle in the namespace
- by reset "security.opendatahub.io/inject-trusted-ca-bundle" to "true" does not get CM generated again (i think this is ok as long as we have it stated in the documentation) 
- after removing "security.opendatahub.io/inject-trusted-ca-bundle" namespace not get CM generated => **this need double check**
-  the initial installation, did not get CM created into any namespace, => **this need to double verify**
- create new namespace
   1. match reserved name, CM not get created => correct
   2. does not match reserved name, CM gets created => correct
 - manually update CM data, cet gets reconciled back to original bundle => correct
 - update customCABundle value, CM data did not get updated, no log on controllers.CertConfigmapGenerator => **need double verify this**


P.S:
do we need to have something in the "docs" to ask user if they want to enable this feature to have self-signed cert, then should check existing namespace?
e.g if we do not have annotation, see it as a  "true" or "false" ?  (current code  see not found as keep CM  but not gonna create one) => this could related to `NamespaceCreatedPredicate` on update by return from `HasCABundleAnnotationDisabled`
P.S2:
since we watch CM with this name "odh-trusted-ca-bundle", this will be a reserve CM which user should avoid create CM with the same name. if user has this CM already, it will be injected as data.odh-data-bundle.crt to it

VaishnaviHire commented:
> part still now work:
> 
> * when  "security.opendatahub.io/inject-trusted-ca-bundle:false"  set back to "security.opendatahub.io/inject-trusted-ca-bundle:true"
> * when remove ""security.opendatahub.io/inject-trusted-ca-bundle: false/true"
>   above 2 not get CM created even DSCI has  `.spec.trustedCABundle.managementState` to Managed
> 

The latest update should fix this. Fixed the filtering for [namespace update events](https://github.com/opendatahub-io/opendatahub-operator/pull/848/files#diff-8b547a2bf349496673680e625159df2d79b818ec9c7eb818722d0d7b87943d11R133).

harshad16 commented:
> as it's currently created by and used by odh notebook controller?
> 
> If yes, then this new feature is awesome, as it adds cross-component support for that odh configmap. In all namespaces, I assume, also data science project workbench namespaces?
> 
> Are you planning to do this for odh notebook controller in this new way here, too?

@shalberd
yes, this would be work on each components (notebook-controller)
we would be using this config map mount methodology.
instead of looking up the proxy as there is in the existing code of the notebook-controller.


about the getting these configmap on all the DS project, is being planned and worked out by the team
it would be get these as well.

etirelli reviewed (approved):
lgtm

shalberd commented:
@harshad16 @VaishnaviHire 

"we would be using this config map mount methodology.
instead of looking up the proxy as there is in the existing code of the notebook-controller."

just be aware that for https://docs.openshift.com/container-platform/4.12/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki to work, that is, for [ca-bundle.crt](https://github.com/opendatahub-io/opendatahub-operator/pull/848/files#diff-5c3ce1dbaa58511a40befd289a83c00935e6b60d4ff5ed47cf1dbb82e73a58bfR48) to be populated with actual values, the proxy at cluster central level probably needs to be active as well.

In odh notebook controller, you only added the old configmap when proxy was active. So it'd be interesting of that configmap mechanism and value of ca-bundle.crt is there when Proxy is not active.
